RFC: libutils libutee etc. as shared libraries

jforissier opened the pull request:
This RFC introduces a config flag to build user-mode libraries (libutee, libutils etc.) as shared objects.
I wrote this patch set when I was implementing the dynamic link support in the ELF loader, because it was quite convenient to test the feature. But, without the ability to share code and RO data between TAs, it was quite useless in practice.

Now that @jenswi-linaro has proposed #2801 "Shared read-only pages between TAs", things are different and thus I'm pushing this as an RFC. With both features combined we can probably save significant amounts of memory and load many more TAs simultaneously. 

FYI, here is the output of `xtest 1019` (calling `TEE_Panic()` in a library) filtered through `symbolize.py`, with all libraries built as shared objects. Pretty cool, no? ;-)

```
E/TC:? 0 TA panicked with code 0x0
E/TC:? 0 Status of TA 5b9e0e40-2636-11e1-ad9e-0002a5d5c51b (0xe171b10) (active)
E/TC:? 0  arch: arm  load address: 0x105000 ctx-idr: 2
E/TC:? 0  stack: 0x102000 10240
E/TC:? 0  region 0: va 0x100000 pa 0xe100000 size 0x1000 flags ---R-X
E/TC:? 0  region 1: va 0x102000 pa 0xe3f0000 size 0x3000 flags rw-RW-
E/TC:? 0  region 2: va 0x105000 pa 0xe308000 size 0x6000 flags r-x--- [0] .ta_head .text .plt .rodata .ARM.extab .ARM.exidx .got .dynsym .rel.plt .dynamic .dynstr .hash .rel.dyn
E/TC:? 0  region 3: va 0x10b000 pa 0xe30e000 size 0xe2000 flags rw---- [0] .data .bss
E/TC:? 0  region 4: va 0x1ed000 pa 0xe3f3000 size 0x1000 flags r-x--- [1] .hash .dynsym .dynstr .rel.plt .plt .text .ARM.extab .ARM.exidx
E/TC:? 0  region 5: va 0x1fd000 pa 0xe403000 size 0x1000 flags rw---- [1] .dynamic .got
E/TC:? 0  region 6: va 0x1fe000 pa 0xe482000 size 0xe000 flags r-x--- [2] .hash .dynsym .dynstr .rel.dyn .rel.plt .plt .text .rodata .ARM.extab .ARM.exidx
E/TC:? 0  region 7: va 0x21b000 pa 0xe49f000 size 0x1000 flags rw---- [2] .dynamic .got .data .bss
E/TC:? 0  region 8: va 0x21c000 pa 0xe43f000 size 0xd000 flags r-x--- [3] .hash .dynsym .dynstr .rel.dyn .rel.plt .plt .text .rodata .ARM.extab .ARM.exidx
E/TC:? 0  region 9: va 0x239000 pa 0xe45c000 size 0x3000 flags rw---- [3] .data.rel.ro .dynamic .got .data .bss
E/TC:? 0  region 10: va 0x23c000 pa 0xe54f000 size 0x38000 flags r-x--- [4] .hash .dynsym .dynstr .rel.dyn .rel.plt .plt .text .rodata .ARM.extab .ARM.exidx
E/TC:? 0  region 11: va 0x283000 pa 0xe596000 size 0x4000 flags rw---- [4] .data.rel.ro .dynamic .got .data .bss
E/TC:? 0  region 12: va 0x287000 pa 0xe300000 size 0x3000 flags r-----
E/TC:? 0  [0] 5b9e0e40-2636-11e1-ad9e-0002a5d5c51b @ 0x105000 (optee_test/out/ta/os_test/5b9e0e40-2636-11e1-ad9e-0002a5d5c51b.elf)
E/TC:? 0  [1] ffd2bded-ab7d-4988-95ee-e4962fff7154 @ 0x1ed000 (optee_test/out/ta/os_test_lib/libos_test.so)
E/TC:? 0  [2] 71855bba-6055-4293-a63f-b0963a737360 @ 0x1fe000 (optee_os/out/arm/ta_arm32-lib/libutils/libutils.so)
E/TC:? 0  [3] 527f1a47-b92c-4a74-95bd-72f19f4a6f74 @ 0x21c000 (optee_os/out/arm/ta_arm32-lib/libutee/libutee.so)
E/TC:? 0  [4] 87bb6ae8-4b1d-49fe-9986-2b966132c309 @ 0x23c000 (optee_os/out/arm/ta_arm32-lib/libmbedtls/libmbedtls.so)
E/TC:? 0 Call stack:
E/TC:? 0  0x00227760 utee_panic at optee_os/lib/libutee/arch/arm/utee_syscalls_a32.S:52
E/TC:? 0  0x00225105 TEE_Panic at optee_os/lib/libutee/tee_api_panic.c:13
E/TC:? 0  0x001ed317 os_test_shlib_panic at optee_test/ta/os_test_lib/os_test_lib.c:17
E/TC:? 0  0x00107c8b ta_entry_call_lib_panic at optee_test/ta/os_test/os_test.c:1089
E/TC:? 0  0x00107df1 TA_InvokeCommandEntryPoint at optee_test/ta/os_test/ta_entry.c:116
E/TC:? 0  0x002276bf entry_invoke_command at optee_os/lib/libutee/arch/arm/user_ta_entry.c:191
E/TC:? 0  0x0022771b __utee_entry at optee_os/lib/libutee/arch/arm/user_ta_entry.c:219
```

jforissier commented:
Bug: in-tree early TAs probably need to be linked with `-static` when shared libs are enabled. Or, alternatively, dependent library need to be installed as "early libraries" (more complex).
I need to fix that.

jenswi-linaro commented on ta/ta.mk:
```diff
@@ -48,28 +48,45 @@ base-prefix := $(sm)-
 libdir = lib/libmpa
-include mk/lib.mk
+ifeq ($(CFG_ULIBS_SHARED),y)
+libuuid = 39b498d9-1e1f-4ae0-a9e1-6d1caf8ec731
```
Wouldn't it make sense to have different uuids for mbedtls and mpa?

jforissier replied:
Huh, yes, they are different already!?

jenswi-linaro replied:
Sorry, my mistake.

jenswi-linaro commented:
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

jforissier commented:
Update. Will link in-tree TAs statically, always.

jenswi-linaro commented:
LGTM

jforissier commented:
Sorry, one more fixup... I forgot to strip the libraries, so `xtest 6016` was unhappy (out of memory).

There is some duplication between `mk/lib.mk` and `ta/arch/arm/link_shlib.mk` which I am not particularly proud of :( but the variable names are quite different so would it really be worth trying to unify?

etienne-lms commented on ta/ta.mk:
```diff
@@ -48,28 +48,45 @@ base-prefix := $(sm)-
 libdir = lib/libutils
+ifeq ($(CFG_ULIBS_SHARED),y)
+libuuid = 71855bba-6055-4293-a63f-b0963a737360
+endif
```
These are ok but could maybe be centralized in a `lib/uuids.mk` file, a set from a library makefile script.
`ta/ta.mk` looks not the right place.

jforissier replied:
IMO it is the right place for defining the UUID, because it is where we define which user-mode libraries we want to build and include in the TA dev kit, their name, their source directory, and then logically their UUID.
In fact I think the name of the makefile (`ta/ta.mk`) is not very good, especially now that we also have in-tree TAs.

How about I move the `ifeq ($(CFG_ULIBS_SHARED),y)` to `mk/lib.mk` instead of repeating it for each library?

jenswi-linaro commented:
It's not entirely clear to my why this duplication is needed. It would be nice to be rid of it, or if it's too much trouble to do right now at least a comment in the makefiles explaining the situation.

jforissier commented on mk/lib.mk:
```diff
@@ -1,14 +1,19 @@
+# libuuid       the UUID of the lib (optional, will produced a shared library)
+# libl		other libraries this library depends on; used to generate the
+#               proper link arguments (-Lxxx -lyyy) and to add dependencies
+#               on the needed .so files (optional, only libuuid is set)
```
Will rework this.

jforissier commented on mk/lib.mk:
```diff
@@ -20,13 +25,26 @@ endif
+lib-needed-so-files := $(foreach l,$(libl),$(lib-shlibfile-$(l)))
+lib-Ll-args := $(foreach l,$(libl),-L$(lib-libdir-$(l)) -l$(l))
+endif
+cleanfiles	:= $(lib-libfile) $(lib-shlibfile) $(lib-shlibtafile) $(lib-libuuidln) $(cleanfiles)
```
Will fix

jforissier commented on mk/lib.mk:
```diff
@@ -20,13 +25,26 @@ endif
+lib-Ll-args := $(foreach l,$(libl),-L$(lib-libdir-$(l)) -l$(l))
+endif
+cleanfiles	:= $(lib-libfile) $(lib-shlibfile) $(lib-shlibtafile) $(lib-libuuidln) $(cleanfiles)
+libfiles	:= $(lib-libfile) $(lib-shlibfile) $(lib-shlibtafile) $(lib-libuuidln) $(libfiles)
```
Yes, not strictly necessary I suppose but it should logically be there. Will fix.

jforissier commented on mk/lib.mk:
```diff
@@ -48,4 +86,8 @@ $(objs): $(conf-file)
 lib-use-ld	:=
+lib-shlibfile	:=
+lib-shlibtafile	:=
+lib-libuuidln	:=
```
Yep

jforissier commented:
@jenswi-linaro the duplication is not really needed, but this is the situation we have for static libs already.
- `mk/lib.mk` has rules to generate static libs when building TEE core. `ta/mk/ta_dev_kit.mk` has similar rules to be used when building outside optee_os. Note that when this capability was added, `mk/lib.mk` was *not* refactored (it was a very small duplication however).
- Then shared library support was added to the TA dev kit. So, `ta/mk/ta_dev_kit.mk` was modified and a helper file was written (`ta/arch/arm/link_shlib.mk`).
- Now I am adding support for generating shlibs in the optee_os build. Re-using `ta/arch/arm/link_shlib.mk` would be good but not trivial and then why would we still have duplicate AR rules for in-tree and out-of-tree.

Perhaps this re-work should be handled in a separate patch but I'm not sure it's even worth it. It's not like the compile flags or tons of stuff subject to change are duplicated. These are pretty much static rules for how to invoke the linker to produce a shared lib.

jenswi-linaro commented:
It's fine to save the re-work for another pull request. However, I think it's worth doing eventually.

jforissier commented:
All comments hopefully addressed. I have reverted "user_ta_header.c: use wrapper function for __utee_entry()" and implemented a fix in the ELF loader instead.
Please check commit "core: elf_load: R_ARM_ABS32/R_AARCH64_ABS64 relocations against undefined symbol".

jenswi-linaro commented:
LGTM. Nice fix with relocation.

jforissier commented:
> LGTM. Nice fix with relocation.

Thanks. It looks like we will need a wrapper anyways when merging this with your "shared RO pages" series, because the entry point supplied to the linker cannot be an undefined symbol (I'm talking about the `-e__utee_entry` flag which should become `-e__ta_entry`, a wrapper for `__utee_entry`)
